zcash_keys/src/keys/transparent.rs

```diff
@@ -0,0 +1,340 @@
+            return Err(());
+        }
+        let seckey = &seckey[1..];
+
+        //    /* sequence length constructor */
+        //    if (end - seckey < 1 || !(*seckey & 0x80u)) {
+        //        return 0;
+        //    }
+        //    size_t lenb = *seckey & ~0x80u; seckey++;
+        //    if (lenb < 1 || lenb > 2) {
+        //        return 0;
+        //    }
+        //    if (end - seckey < lenb) {
+        //        return 0;
+        //    }
+        if seckey.is_empty() || (seckey[0] & 0x80 == 0) {
+            return Err(());
+        }
+        let lenb = usize::from(seckey[0] & !0x80);
+        let seckey = &seckey[1..];
+        if !(1..=2).contains(&lenb) {
+            return Err(());
+        }
+        if seckey.len() < lenb {
+            return Err(());
+        }
+
+        //    /* sequence length */
+        //    size_t len = seckey[lenb-1] | (lenb > 1 ? seckey[lenb-2] << 8 : 0u);
+        //    seckey += lenb;
+        //    if (end - seckey < len) {
+        //        return 0;
+        //    }
+        let len_low_bits = usize::from(seckey[lenb - 1]);
+        let len_high_bits = if lenb > 1 {
+            usize::from(seckey[lenb - 2]) << 8
+        } else {
+            0
+        };
+        let len = len_low_bits | len_high_bits;
+        let seckey = &seckey[lenb..];
+        if seckey.len() < len {
+            return Err(());
+        }
+
+        //    /* sequence element 0: version number (=1) */
+        //    if (end - seckey < 3 || seckey[0] != 0x02u || seckey[1] != 0x01u || seckey[2] != 0x01u) {
+        //        return 0;
+        //    }
+        //    seckey += 3;
+        if seckey.len() < 3 || seckey[0] != 0x02 || seckey[1] != 0x01 || seckey[2] != 0x01 {
+            return Err(());
+        }
+        let seckey = &seckey[3..];
+
+        //    /* sequence element 1: octet string, up to 32 bytes */
+        //    if (end - seckey < 2 || seckey[0] != 0x04u) {
+        //        return 0;
+        //    }
+        //    size_t oslen = seckey[1];
+        //    seckey += 2;
+        //    if (oslen > 32 || end - seckey < oslen) {
+        //        return 0;
+        //    }
+        //    memcpy(out32 + (32 - oslen), seckey, oslen);
+        //    if (!secp256k1_ec_seckey_verify(ctx, out32)) {
+        //        memset(out32, 0, 32);
+        //        return 0;
+        //    }
+        //    return 1;
+        //}
+        if seckey.len() < 2 || seckey[0] != 0x04 {
+            return Err(());
+        }
+        let oslen = usize::from(seckey[1]);
+        let seckey = &seckey[2..];
+        if oslen > 32 || seckey.len() < oslen {
+            return Err(());
+        }
+        let mut secret_buf = [0u8; 32];
+        secret_buf[(32 - oslen)..].copy_from_slice(&seckey[..oslen]);
+        match SecretKey::from_bytes(&secret_buf) {
+            Ok(secret) => Ok(Self { secret, compressed }),
+            Err(_) => {
+                secret_buf.zeroize();
```

Comment:
We want to zeroize this unconditionally. In the C++ code `out32` was how the secret was returned in the success case; here `secret_buf` is internal.